crates/core_simd/src/vector.rs

```diff
@@ -12,21 +9,394 @@ pub use uint::*;
+where
+    LaneCount<LANES>: SupportedLaneCount,
+    Element: SimdElement + PartialOrd,
+{
+    #[inline]
+    fn partial_cmp(&self, other: &Self) -> Option<core::cmp::Ordering> {
+        // TODO use SIMD equality
+        self.to_array().partial_cmp(other.as_ref())
+    }
+}
+
+impl<Element, const LANES: usize> Eq for Simd<Element, LANES>
+where
+    LaneCount<LANES>: SupportedLaneCount,
+    Element: SimdElement + Eq,
+{
+}
+
+impl<Element, const LANES: usize> Ord for Simd<Element, LANES>
+where
+    LaneCount<LANES>: SupportedLaneCount,
+    Element: SimdElement + Ord,
+{
+    #[inline]
+    fn cmp(&self, other: &Self) -> core::cmp::Ordering {
+        // TODO use SIMD equality
+        self.to_array().cmp(other.as_ref())
+    }
+}
+
+impl<Element, const LANES: usize> core::hash::Hash for Simd<Element, LANES>
+where
+    LaneCount<LANES>: SupportedLaneCount,
+    Element: SimdElement + core::hash::Hash,
+{
+    #[inline]
+    fn hash<H>(&self, state: &mut H)
+    where
+        H: core::hash::Hasher,
+    {
+        self.as_array().hash(state)
+    }
+}
+
+// array references
+impl<Element, const LANES: usize> AsRef<[Element; LANES]> for Simd<Element, LANES>
+where
+    LaneCount<LANES>: SupportedLaneCount,
+    Element: SimdElement,
+{
+    #[inline]
+    fn as_ref(&self) -> &[Element; LANES] {
+        &self.0
+    }
+}
+
+impl<Element, const LANES: usize> AsMut<[Element; LANES]> for Simd<Element, LANES>
+where
+    LaneCount<LANES>: SupportedLaneCount,
+    Element: SimdElement,
+{
+    #[inline]
+    fn as_mut(&mut self) -> &mut [Element; LANES] {
+        &mut self.0
+    }
+}
+
+// slice references
+impl<Element, const LANES: usize> AsRef<[Element]> for Simd<Element, LANES>
+where
+    LaneCount<LANES>: SupportedLaneCount,
+    Element: SimdElement,
+{
+    #[inline]
+    fn as_ref(&self) -> &[Element] {
+        &self.0
+    }
+}
+
+impl<Element, const LANES: usize> AsMut<[Element]> for Simd<Element, LANES>
+where
+    LaneCount<LANES>: SupportedLaneCount,
+    Element: SimdElement,
+{
+    #[inline]
+    fn as_mut(&mut self) -> &mut [Element] {
+        &mut self.0
+    }
+}
+
+// vector/array conversion
+impl<Element, const LANES: usize> From<[Element; LANES]> for Simd<Element, LANES>
+where
+    LaneCount<LANES>: SupportedLaneCount,
+    Element: SimdElement,
+{
+    fn from(array: [Element; LANES]) -> Self {
+        Self(array)
+    }
+}
+
+impl<Element, const LANES: usize> From<Simd<Element, LANES>> for [Element; LANES]
+where
+    LaneCount<LANES>: SupportedLaneCount,
+    Element: SimdElement,
+{
+    fn from(vector: Simd<Element, LANES>) -> Self {
+        vector.to_array()
+    }
+}
+
 mod sealed {
     pub trait Sealed {}
 }
+use sealed::Sealed;
+
+/// Marker trait for types that may be used as SIMD vector elements.
+pub unsafe trait SimdElement: Sealed + Copy {
+    /// The mask element type corresponding to this element type.
+    type Mask: MaskElement;
+}
```

Comment:
Why is this unsafe to implement?

Reply:
probably makes rustc flip out and ICE if you try to do simd over a non-primitive, if i had to guess.

Reply:
Yeah, it's unsafe for two reasons--one is that rustc will produce unfortunate monomorphisation errors for non-scalars.  But another reason to make it unsafe is in the future `repr(simd)` could be expanded to some arbitrary type that we haven't considered, and it may suddenly allow some broken things.

Reply:
This is fair, I basically just wanted to get "presumed unsafe to implement because it relies on compiler-backed monomorphization guarantees" on the record. :^)
```suggestion
/// Marker trait for types that may be used as SIMD vector elements.
/// SAFETY: This trait implies the compiler knows how to monomorphize the element in question.
/// Implementing it for a type asserts the compiler won't miscompile it in a `#[repr(simd)]` struct.
pub unsafe trait SimdElement: Sealed + Copy {
    /// The mask element type corresponding to this element type.
    type Mask: MaskElement;
}
```